Applications/FileIO/SHPInterface.cpp

```diff
@@ -213,87 +213,109 @@ void SHPInterface::readPolygons(const SHPHandle& hSHP, int numberOfElements,
+        {
+            std::vector<int> const& vec = *properties.getPropertyVector<int>(name);
+            int const field = DBFAddField(hDBF, name.c_str(), FTInteger, 16, 0);
+            for (int i = 0; i < n_recs; ++i)
             {
-                auto const* const materialIds =
-                    mesh.getProperties().getPropertyVector<int>("MaterialIDs");
-                DBFWriteIntegerAttribute(hDBF, polygon_id, mat_field, (*materialIds)[i]);
+                std::size_t const elem_idx = DBFReadIntegerAttribute(hDBF, i, elem_id_field);
+                DBFWriteIntegerAttribute(hDBF, i, field, vec[elem_idx]);
             }
-            unsigned nNodes (e->getNumberOfBaseNodes());
-            padfX = new double[nNodes+1];
-            padfY = new double[nNodes+1];
-            padfZ = new double[nNodes+1];
-            for (unsigned j=0; j<nNodes; ++j)
+        }
+        else if (properties.existsPropertyVector<double>(name))
+        {
+            std::vector<double> const& vec = *properties.getPropertyVector<double>(name);
+            int const field = DBFAddField(hDBF, name.c_str(), FTDouble, 33, 16);
+            for (int i = 0; i < n_recs; ++i)
             {
-                padfX[j]=(*e->getNode(j))[0];
-                padfY[j]=(*e->getNode(j))[1];
-                padfZ[j]=(*e->getNode(j))[2];
+                std::size_t const elem_idx = DBFReadIntegerAttribute(hDBF, i, elem_id_field);
+                DBFWriteDoubleAttribute(hDBF, i, field, vec[elem_idx]);
             }
-            // Last node == first node to close the polygon
-            padfX[nNodes]=(*e->getNode(0))[0];
-            padfY[nNodes]=(*e->getNode(0))[1];
-            padfZ[nNodes]=(*e->getNode(0))[2];
-            // write the first three node ids to the dbf-file (this also specifies a QUAD uniquely)
-            DBFWriteIntegerAttribute(hDBF, polygon_id, node0_field, e->getNode(0)->getID());
-            DBFWriteIntegerAttribute(hDBF, polygon_id, node1_field, e->getNode(1)->getID());
-            DBFWriteIntegerAttribute(hDBF, polygon_id, node2_field, e->getNode(2)->getID());
-
-            SHPObject* object =
-                SHPCreateObject(SHPT_POLYGON, polygon_id++, 0, nullptr, nullptr,
-                                ++nNodes, padfX, padfY, padfZ, nullptr);
-            SHPWriteObject(hSHP, -1, object);
-
-            // Note: cleaning up the coordinate arrays padfX, -Y, -Z results in a crash, I assume that shapelib removes them
-            delete object;
         }
     }
-
-    SHPClose(hSHP);
     DBFClose(hDBF);
-    INFO("Shape export of 2D mesh '%s' successful.", mesh.getName().c_str());
-
+    INFO("Shape export of 2D mesh '%s' finished.", mesh.getName().c_str());
     return true;
 }
 
+SHPObject* SHPInterface::createShapeObject(MeshLib::Element const& e,
+                                           std::size_t const shp_record)
+{
+    unsigned const nNodes(e.getNumberOfBaseNodes());
+    double* padfX = new double[nNodes + 1];
```

Comment:
New memory allocation - who deletes it?

Reply:
Just now I saw the removed comment (previous line 287) that says that shapelib crashes when deleting the memory. Please, insert the comment somewhere in the method.